core/src/block.rs

```diff
@@ -19,894 +19,614 @@ use iroha_data_model::{block::*, events::prelude::*, transaction::prelude::*};
+}
 
-        pub fn hash(&self) -> HashOf<Self> {
-            HashOf::new(&self.header).transmute()
+impl Auditable for PendingBlock {
+    /// Revalidate a block against the current state of the world.
+    ///
+    /// # Errors
+    /// - Block is empty
+    /// - Block has committed transactions
+    /// - There is a mismatch between candidate block height and actual blockchain height
+    /// - There is a mismatch between candidate block previous block hash and actual latest block hash
+    /// - Block header transaction hashes don't match with computed transaction hashes
+    /// - Error during revalidation of individual transactions
+    #[allow(clippy::too_many_lines)]
+    fn revalidate<const IS_GENESIS: bool>(
+        self,
+        transaction_validator: &TransactionValidator,
+        wsv: &WorldStateView,
+        latest_block: Option<HashOf<VersionedCommittedBlock>>,
+        block_height: u64,
+    ) -> Result<Self, eyre::Report> {
+        if self.transactions.is_empty() && self.rejected_transactions.is_empty() {
+            bail!("Block is empty");
+        }
+
+        if self.has_committed_transactions(wsv) {
+            bail!("Block has committed transactions");
+        }
+
+        if latest_block != self.header.previous_block_hash {
+            bail!(
+                    "Mismatch between the actual and expected hashes of the latest block. Expected: {:?}, actual: {:?}",
+                    latest_block,
+                    &self.header.previous_block_hash
+                );
         }
 
-        /// Return signatures that are verified with the `hash` of this block,
-        /// removing all other signatures.
-        #[inline]
-        pub fn retain_verified_signatures(&mut self) -> impl Iterator<Item = &SignatureOf<Self>> {
-            self.signatures.retain_verified_by_hash(self.hash())
+        if block_height + 1 != self.header.height {
+            bail!(
+                    "Mismatch between the actual and expected heights of the block. Expected: {}, actual: {}",
+                    block_height + 1,
+                    self.header.height
+                );
         }
 
-        /// Commit block to the store.
-        /// When calling this function, the user is responsible for the validity of the block signatures.
-        /// Preference should be given to [`Self::commit`], where signature verification is built in.
-        #[inline]
-        pub fn commit_unchecked(self) -> CommittedBlock {
-            let Self {
-                header,
-                rejected_transactions,
-                transactions,
-                event_recommendations,
-                signatures,
-            } = self;
-
-            CommittedBlock {
-                event_recommendations,
-                header,
-                rejected_transactions,
-                transactions,
-                signatures: signatures.transmute(),
-            }
-        }
+        let wsv = wsv.clone();
```

Comment:
Another implicit `wsv` clone.

Reply:
The cloning was introduce by someone other than me in order to solve a bug. This is a immutable reference but it infact has interior mutability. If we don't clone then the wsv is modified and that is incorrect behaviour. We need a local copy we can mutate because we need to validate a series of transactions and how they affect each other.